aquarium/aquarium.js

```diff
@@ -1701,6 +1704,106 @@ function initialize() {
     render(monoProjection, viewInverseTemp);
   }
 
+  function onXRFrame(now, frame) {
+    let session = frame.session;
+    let refSpace = g_xrImmersiveRefSpace;
+    let pose = frame.getViewerPose(refSpace);
+
+    var now = theClock.getTime();
+    var elapsedTime;
+    if(then == 0.0) {
+      elapsedTime = 0.0;
+    } else {
+      elapsedTime = now - then;
+    }
+    then = now;
+
+    if (g.net.sync) {
+      clock = now * g.globals.speed;
+      eyeClock = now * g.globals.eyeSpeed;
+    } else {
+      // we have our own clock.
+      clock += elapsedTime * g.globals.speed;
+      eyeClock += elapsedTime * g.globals.eyeSpeed;
+    }
+
+    frameCount++;
+    g_fpsTimer.update(elapsedTime);
+    fpsElem.innerHTML = g_fpsTimer.averageFPS;
+
+    if (g_shadersNeedUpdate) {
+      setShaders(true);
+      g_shadersNeedUpdate = false;
+    }
+
+    if (g_vrUi) {
+      // Set fps and prepare rendering it.
+      g_vrUi.setFps(g_fpsTimer.averageFPS);
+
+      // Query gamepad button clicked event.
+      g_vrUi.queryGamepadStatus();
+
+      // TODO: Support VRUI when doing multiview rendering.
+      if (!useMultiviewForStereo() && g_vrUi.isMenuMode) {
+
+        // When VR UI in menu mode, UI need a cursor to help user do select operation. Currently, cursor uses
+        // head-neck model which means a point in front of user and user could move the point by rotating their head(with HMD).
+        // A click event will be triggered when user stare at a label 2 seconds.
+        // TODO : add gamepad support to control cursor and trigger select event with VR controllers.
+
+        // Jquery selector description.
+        var selectorDescription;
+
+        // VR UI return whether there is an option been selected in VR mode.
+        var clickedLabel = g_vrUi.queryClickedLabel([0, 0, 0], g_frameData.pose.orientation);
+        if (clickedLabel != null) {
+          if (clickedLabel.isAdvancedSettings) {
+            selectorDescription = "#optionsContainer > div:contains(" + clickedLabel.name + ")";
+            $(selectorDescription).click();
+          } else if (clickedLabel.name == "options") {
+            $("#options").click();
+          } else {
+            selectorDescription = "#setSetting" + clickedLabel.name;
+            $(selectorDescription).click();
+          }
+        }
+      }
+    }
+
+    // Using head-neck model in VR mode because of unclear distance measurement(vr return position using meters),
+    // user could see around but couldn't move around.
+    eyePosition[0] = g.globals.eyeRadius;
+    eyePosition[1] = g.globals.eyeHeight;
+    eyePosition[2] = g.globals.eyeRadius;
+
+    let vrPose = [pose.transform.orientation.x, pose.transform.orientation.y, pose.transform.orientation.z, 1]
```

Comment:
Tried it on Oculus Browser and Wolvic on Quest 2, and it does not work. Going to investigate it.
Tested on cardboard on Galaxy A34 on chrome versions 120.0.6099.145 and .193 .
Previously, I had problems with older versions of Chrome crashing on entering XR, but this does not happen anymore.

I made this change but later removed this code because it was unnecessary since all matrices were coming from xrPose.